Merge ToC Extension into core

marthacryan opened the pull request:
## References

As discussed in [this issue](https://github.com/jupyterlab/team-compass/issues/63) and in the weekly dev meetings, [the ToC extension](https://github.com/jupyterlab/jupyterlab-toc) seems to have enough support to be merged into jupyterlab core. 

## Code changes

To follow the conventions of other core packages, the ToC extension is split into a `toc` package and a `toc-extension` package. There are a few changes to the configurations and imports to follow conventions for packages / make the extension work in core. Otherwise it's importing the toc code "as-is" at the moment. 

## User-facing changes

This will cause the ToC extension to appear as an option in the left side panel along with the File Browser, Running Terminals and Kernels, etc.

ajbozarth commented on packages/toc-extension/package.json:
```diff
@@ -0,0 +1,70 @@
+{
+  "name": "@jupyterlab/toc-extension",
+  "version": "4.0.0",
+  "private": false,
```
I think we can remove this line

ajbozarth commented on packages/toc/.gitignore:
```diff
@@ -0,0 +1,7 @@
+*.bundle.*
```
remove this file

ajbozarth commented on packages/toc/.prettierignore:
```diff
@@ -0,0 +1,2 @@
+node_modules
```
remove this file

ajbozarth commented on packages/toc/.prettierrc:
```diff
@@ -0,0 +1,3 @@
+{
```
remove this file

ajbozarth commented on packages/toc/LICENSE:
```diff
@@ -0,0 +1,27 @@
+Copyright (c) 2017, Project Jupyter Contributors
```
remove this file

ajbozarth commented on packages/toc/README.md:
```diff
@@ -0,0 +1,52 @@
+# jupyterlab-toc
```
The relevant contents of this file should be moved to the docs and then remove this file

ajbozarth commented on packages/toc/binder/postBuild:
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env bash
```
remove this file

ajbozarth commented on packages/toc/binder/requirements.txt:
```diff
@@ -0,0 +1 @@
+jupyterlab>=1.0
```
remove this file

ajbozarth commented on packages/toc/notebooks/demo.ipynb:
```diff
@@ -0,0 +1,994 @@
+{
```
remove this file

ajbozarth commented on packages/toc/package.json:
```diff
@@ -0,0 +1,79 @@
+  "keywords": [
+    "jupyterlab"
+  ],
+  "homepage": "https://github.com/jupyterlab/jupyterlab-toc",
```
same comments as the other `package.json` update urls and description, and remove the private line

ajbozarth commented on packages/toc/tslint.json:
```diff
@@ -0,0 +1,109 @@
+{
```
remove this file

ajbozarth commented on packages/toc/yarn.lock:
```diff
@@ -0,0 +1,3041 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
remove this file

telamonian commented:
Wow. Beatiful job preserving the history of the original project.

This should probably get its own test suite before it goes into core. Are there any tests from the existing extension that can be salvaged here?

ajbozarth commented:
@telamonian all the toc code is in this PR, it didn't have tests in it's repo :(

telamonian commented:
Oh. Then I guess the history thing is slightly less impressive. Though I do still think its nice to be able to pick up everyone's contributions.

But yeah, this needs tests

blink1073 commented:
@jupyterlab/committers let's avoid merging any other PRs until this goes in to avoid huge headaches.

blink1073 commented:
@telamonian I agree we need some tests but I don't want to hold up this particular PR for too long to avoid major merge conflicts.  Worst case we can disable this extension by default if we feel it isn't stable enough for 3.0.

blink1073 commented:
@marthacryan please let us know once you've finished addressing pressing concerns, and we can merge and iterate.  The docs failure is unrelated but there is a lint error (please run `jlpm lint` locally and commit the changes).

marthacryan commented:
Fixed the remaining lint errors - I had to use unsafe react methods (such as `UNSAFE_componentWillReceiveProps`) to fix a couple of these errors, but this method is used in another core extension in Jupyterlab [here](https://github.com/jupyterlab/jupyterlab/blob/master/packages/extensionmanager/src/widget.tsx#L509), so I thought it'd be ok for now - let me know if I should fix that in this PR. 

I also changed the version of `toc` and `toc-extension` to `4.2.0-alpha.0`.

jasongrout commented:
Just checking - did you change the collapsing functionality already to a signal or something, or is that something we'll need to do in another PR?

marthacryan commented:
@jasongrout No I didn't include the collapsing signal in this PR (it's currently a setting that makes the collapsing disabled by default).

blink1073 commented:
I opened https://github.com/jupyterlab/jupyterlab/issues/8546 for tracking.

blink1073 commented:
The usage test is hanging again, merging.  Thanks @marthacryan!  This is an epic first contribution :smile:

marthacryan commented:
@blink1073 Thanks you!
